File: backend/providers/github.py
"""
GitHub Evidence Provider
Retrieves commits, pull requests, and deployment telemetry from GitHub REST API.
Provides graceful demo fallback if credentials are absent or if rate-limited.
"""

import os
import json
import urllib.request
import urllib.error
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from models import (
    EvidenceEvent,
    EvidenceSource,
    ConnectorStatus,
    AcquisitionResult,
    GitHubAcquisitionRequest,
)
from .base import EvidenceProvider
from .demo import DemoProvider


class GitHubProvider(EvidenceProvider):
    """
    Acquires version control and deployment evidence from GitHub.
    Isolated service with strict timeout handling and fallback guarantee.
    """

    GITHUB_API_BASE = "https://api.github.com"
    TIMEOUT_SECONDS = 7

    def __init__(self, token: Optional[str] = None):
        self._explicit_token = token

    @property
    def name(self) -> str:
        return "github"

    def _get_token(self, request_token: Optional[str] = None) -> Optional[str]:
        return request_token or self._explicit_token or os.getenv("GITHUB_TOKEN")

    async def get_status(self, token_override: Optional[str] = None) -> ConnectorStatus:
        token = self._get_token(token_override)
        if not token:
            return ConnectorStatus(
                connected=False,
                available=True,
                error="GITHUB_TOKEN not configured in environment or request",
                fallback_available=True,
                details={"auth_method": "none", "hint": "Provide a PAT to query private or live public repositories"},
            )

        # Quick probe to GitHub API rate limit endpoint
        try:
            req = urllib.request.Request(
                f"{self.GITHUB_API_BASE}/rate_limit",
                headers={
                    "Authorization": f"Bearer {token}",
                    "User-Agent": "BlackBox-Incident-Investigator/1.0",
                    "Accept": "application/vnd.github+json",
                },
            )
            with urllib.request.urlopen(req, timeout=self.TIMEOUT_SECONDS) as resp:
                if resp.status == 200:
                    data = json.loads(resp.read().decode("utf-8"))
                    remaining = data.get("rate", {}).get("remaining", 0)
                    return ConnectorStatus(
                        connected=True,
                        available=True,
                        error=None,
                        fallback_available=True,
                        details={"rate_limit_remaining": remaining, "authenticated": True},
                    )
        except urllib.error.HTTPError as e:
            return ConnectorStatus(
                connected=False,
                available=True,
                error=f"GitHub API returned HTTP {e.code}: {e.reason}",
                fallback_available=True,
                details={"http_code": e.code},
            )
        except Exception as e:
            return ConnectorStatus(
                connected=False,
                available=True,
                error=f"GitHub connectivity check failed: {str(e)}",
                fallback_available=True,
            )

        return ConnectorStatus(connected=False, available=True, fallback_available=True)

    async def collect_evidence(
        self,
        repository: str = "payments-team/payments-api",
        branch: Optional[str] = "main",
        incident_start_time: Optional[str] = None,
        incident_end_time: Optional[str] = None,
        personal_access_token: Optional[str] = None,
        include_pull_requests: bool = True,
        include_deployments: bool = True,
        **kwargs: Any,
    ) -> AcquisitionResult:
        """
        Queries GitHub for commits, PRs, and deployments in the incident window.
        Gracefully falls back to demo data if credentials are missing or API fails.
        """
        token = self._get_token(personal_access_token)
        events: List[EvidenceEvent] = []
        warnings: List[str] = []
        used_fallback = False
        error_message: Optional[str] = None

        if not token:
            used_fallback = True
            error_message = "No GITHUB_TOKEN provided; loaded authentic demo Git evidence."
            events = await self._generate_fallback_events(repository, branch)
        else:
            try:
                # 1. Fetch recent commits
                commits = self._fetch_commits(repository, branch, incident_start_time, token)
                for c in commits:
                    ev = self._normalize_commit(c, repository)
                    if ev:
                        events.append(ev)

                # 2. Optionally fetch closed PRs
                if include_pull_requests:
                    try:
                        prs = self._fetch_pull_requests(repository, token)
                        for pr in prs:
                            ev = self._normalize_pull_request(pr, repository)
                            if ev:
                                events.append(ev)
                    except Exception as e:
                        warnings.append(f"Pull request retrieval notice: {str(e)}")

                # 3. Optionally fetch deployments
                if include_deployments:
                    try:
                        deps = self._fetch_deployments(repository, token)
                        for dep in deps:
                            ev = self._normalize_deployment(dep, repository)
                            if ev:
                                events.append(ev)
                    except Exception as e:
                        warnings.append(f"Deployment info retrieval notice: {str(e)}")

                if not events:
                    warnings.append(f"No commits or PRs found for {repository} in timeframe. Loading baseline demo commits.")
                    events = await self._generate_fallback_events(repository, branch)
                    used_fallback = True

            except urllib.error.HTTPError as e:
                used_fallback = True
                error_message = f"GitHub API error HTTP {e.code} ({e.reason}). Activated fallback demo evidence."
                events = await self._generate_fallback_events(repository, branch)
            except Exception as e:
                used_fallback = True
                error_message = f"GitHub connection error: {str(e)}. Activated fallback demo evidence."
                events = await self._generate_fallback_events(repository, branch)

        status = ConnectorStatus(
            connected=not used_fallback,
            available=True,
            error=error_message,
            fallback_available=True,
            details={"repository": repository, "fallback_used": used_fallback},
        )

        return AcquisitionResult(
            source="github",
            events_count=len(events),
            sources_count=1,
            summary_message=f"Collected {len(events)} evidence events from 1 source (GitHub).",
            events=events,
            connector_status=status,
            warnings=warnings,
        )

    def _fetch_commits(
        self, repo: str, branch: Optional[str], since: Optional[str], token: str
    ) -> List[Dict[str, Any]]:
        url = f"{self.GITHUB_API_BASE}/repos/{repo}/commits?per_page=15"
        if branch:
            url += f"&sha={branch}"
        if since:
            url += f"&since={since}"

        req = urllib.request.Request(
            url,
            headers={
                "Authorization": f"Bearer {token}",
                "User-Agent": "BlackBox-Incident-Investigator/1.0",
                "Accept": "application/vnd.github+json",
            },
        )
        with urllib.request.urlopen(req, timeout=self.TIMEOUT_SECONDS) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _fetch_pull_requests(self, repo: str, token: str) -> List[Dict[str, Any]]:
        url = f"{self.GITHUB_API_BASE}/repos/{repo}/pulls?state=closed&sort=updated&per_page=5"
        req = urllib.request.Request(
            url,
            headers={
                "Authorization": f"Bearer {token}",
                "User-Agent": "BlackBox-Incident-Investigator/1.0",
                "Accept": "application/vnd.github+json",
            },
        )
        with urllib.request.urlopen(req, timeout=self.TIMEOUT_SECONDS) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _fetch_deployments(self, repo: str, token: str) -> List[Dict[str, Any]]:
        url = f"{self.GITHUB_API_BASE}/repos/{repo}/deployments?per_page=5"
        req = urllib.request.Request(
            url,
            headers={
                "Authorization": f"Bearer {token}",
                "User-Agent": "BlackBox-Incident-Investigator/1.0",
                "Accept": "application/vnd.github+json",
            },
        )
        with urllib.request.urlopen(req, timeout=self.TIMEOUT_SECONDS) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _normalize_commit(self, raw: Dict[str, Any], repo: str) -> Optional[EvidenceEvent]:
        sha = raw.get("sha", "")
        commit = raw.get("commit", {})
        message = commit.get("message", "Commit")
        headline = message.split("\n")[0]
        author = commit.get("author", {}).get("name") or raw.get("author", {}).get("login", "Unknown")
        timestamp = commit.get("author", {}).get("date") or datetime.now(timezone.utc).isoformat()
        html_url = raw.get("html_url")

        # Heuristic severity check for suspicious changes
        is_suspicious = any(
            k in headline.lower()
            for k in ["pool", "leak", "revert", "hotfix", "timeout", "perf", "cache", "concurrency", "urgent"]
        )
        severity = "warning" if is_suspicious else "info"

        return EvidenceEvent(
            id=f"evt-gh-{sha[:7]}",
            source=EvidenceSource.GITHUB,
            event_type="commit",
            timestamp=timestamp,
            title=f"Commit [{sha[:7]}]: {headline}",
            message=message,
            summary=f"Git commit to {repo} by {author}: '{headline}'",
            severity=severity,
            metadata={
                "sha": sha,
                "author": author,
                "repository": repo,
                "commit_url": html_url,
                "files_changed_count": len(raw.get("files", [])),
            },
            related_identifiers=[sha[:7], author, repo],
            raw_reference=html_url,
            is_root_cause_candidate=is_suspicious,
        )

    def _normalize_pull_request(self, raw: Dict[str, Any], repo: str) -> Optional[EvidenceEvent]:
        pr_number = raw.get("number")
        title = raw.get("title", "")
        body = raw.get("body") or "No description provided."
        merged_at = raw.get("merged_at")
        if not merged_at:
            return None

        author = raw.get("user", {}).get("login", "unknown")
        html_url = raw.get("html_url")

        is_suspicious = any(
            k in title.lower() for k in ["sync", "pool", "refactor", "bulk", "migration", "concurrency", "worker"]
        )

        return EvidenceEvent(
            id=f"evt-gh-pr-{pr_number}",
            source=EvidenceSource.GITHUB,
            event_type="pull_request",
            timestamp=merged_at,
            title=f"PR #{pr_number} Merged: '{title}'",
            message=body,
            summary=f"Pull Request #{pr_number} by @{author} merged into main for {repo}.",
            severity="warning" if is_suspicious else "info",
            metadata={
                "pr_number": pr_number,
                "author": author,
                "merged_at": merged_at,
                "repository": repo,
                "pr_url": html_url,
            },
            related_identifiers=[f"PR#{pr_number}", f"@{author}", repo],
            raw_reference=html_url,
            is_root_cause_candidate=is_suspicious,
        )

    def _normalize_deployment(self, raw: Dict[str, Any], repo: str) -> Optional[EvidenceEvent]:
        dep_id = raw.get("id")
        created_at = raw.get("created_at") or datetime.now(timezone.utc).isoformat()
        env = raw.get("environment", "production")
        ref = raw.get("ref", "main")
        creator = raw.get("creator", {}).get("login", "ci-bot")

        return EvidenceEvent(
            id=f"evt-gh-deploy-{dep_id}",
            source=EvidenceSource.DEPLOYMENT,
            event_type="deployment",
            timestamp=created_at,
            title=f"GitHub Deployment to {env} ({ref})",
            message=f"Automated deployment triggered by {creator} for ref {ref}.",
            summary=f"Deployment {dep_id} promoted to {env} for repository {repo}.",
            severity="info",
            metadata={
                "deployment_id": dep_id,
                "environment": env,
                "ref": ref,
                "creator": creator,
                "repository": repo,
            },
            related_identifiers=[str(dep_id), env, ref],
            raw_reference=raw.get("url"),
            is_root_cause_candidate=False,
        )

    async def _generate_fallback_events(self, repo: str, branch: Optional[str]) -> List[EvidenceEvent]:
        """Loads realistic demo Git events matching the requested repo context."""
        demo_provider = DemoProvider()
        res = await demo_provider.collect_evidence(incident_id="inc-db-pool-exhaustion", filter_source="github")
        events = res.events
        # Add metadata indicating demo fallback
        for e in events:
            e.metadata["is_fallback"] = True
            e.metadata["fallback_repository"] = repo
        return events

File: src/services/providers/github.ts
import {
  EvidenceEvent,
  ConnectorStatus,
  AcquisitionResult,
  GitHubAcquisitionRequest,
} from '../../types/incident';
import { EvidenceProvider } from './base';
import { DemoProvider } from './demo';

export class GitHubProvider implements EvidenceProvider {
  name = 'github';
  private apiBase = 'https://api.github.com';
  private timeoutMs = 7000;

  private getToken(requestToken?: string): string | undefined {
    return requestToken || (typeof process !== 'undefined' ? process.env?.GITHUB_TOKEN : undefined);
  }

  async getStatus(tokenOverride?: string): Promise<ConnectorStatus> {
    const token = this.getToken(tokenOverride);
    if (!token) {
      return {
        connected: false,
        available: true,
        error: 'GITHUB_TOKEN not configured in environment or request',
        fallback_available: true,
        details: {
          auth_method: 'none',
          hint: 'Provide a PAT to query private or live public repositories',
        },
      };
    }

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

      const resp = await fetch(`${this.apiBase}/rate_limit`, {
        headers: {
          Authorization: `Bearer ${token}`,
          'User-Agent': 'BlackBox-Incident-Investigator/1.0',
          Accept: 'application/vnd.github+json',
        },
        signal: controller.signal,
      });
      clearTimeout(timeout);

      if (resp.ok) {
        const data = await resp.json();
        const remaining = data?.rate?.remaining ?? 0;
        return {
          connected: true,
          available: true,
          error: null,
          fallback_available: true,
          details: { rate_limit_remaining: remaining, authenticated: true },
        };
      } else {
        return {
          connected: false,
          available: true,
          error: `GitHub API returned HTTP ${resp.status}: ${resp.statusText}`,
          fallback_available: true,
          details: { http_code: resp.status },
        };
      }
    } catch (err: any) {
      return {
        connected: false,
        available: true,
        error: `GitHub probe failed: ${err.message || String(err)}`,
        fallback_available: true,
      };
    }
  }

  async collectEvidence(params: GitHubAcquisitionRequest): Promise<AcquisitionResult> {
    const repo = params.repository || 'payments-team/payments-api';
    const branch = params.branch || 'main';
    const token = this.getToken(params.personal_access_token);

    const events: EvidenceEvent[] = [];
    const warnings: string[] = [];
    let usedFallback = false;
    let errorMessage: string | null = null;

    if (!token) {
      usedFallback = true;
      errorMessage = 'No GITHUB_TOKEN configured; loaded authentic demo Git evidence.';
      const fallbackEvents = await this.generateFallbackEvents(repo);
      events.push(...fallbackEvents);
    } else {
      try {
        // 1. Fetch commits
        const commits = await this.fetchCommits(repo, branch, params.incident_start_time, token);
        for (const c of commits) {
          const ev = this.normalizeCommit(c, repo);
          if (ev) events.push(ev);
        }

        // 2. Fetch PRs if requested
        if (params.include_pull_requests !== false) {
          try {
            const prs = await this.fetchPullRequests(repo, token);
            for (const pr of prs) {
              const ev = this.normalizePullRequest(pr, repo);
              if (ev) events.push(ev);
            }
          } catch (e: any) {
            warnings.push(`Pull request retrieval notice: ${e.message || String(e)}`);
          }
        }

        // 3. Fetch deployments if requested
        if (params.include_deployments !== false) {
          try {
            const deps = await this.fetchDeployments(repo, token);
            for (const dep of deps) {
              const ev = this.normalizeDeployment(dep, repo);
              if (ev) events.push(ev);
            }
          } catch (e: any) {
            warnings.push(`Deployment retrieval notice: ${e.message || String(e)}`);
          }
        }

        if (events.length === 0) {
          warnings.push(`No commits or PRs found for ${repo}. Activated baseline fallback demo commits.`);
          usedFallback = true;
          const fallbackEvents = await this.generateFallbackEvents(repo);
          events.push(...fallbackEvents);
        }
      } catch (err: any) {
        usedFallback = true;
        errorMessage = `GitHub API request failed (${err.message || String(err)}). Activated fallback demo evidence.`;
        const fallbackEvents = await this.generateFallbackEvents(repo);
        events.push(...fallbackEvents);
      }
    }

    const status: ConnectorStatus = {
      connected: !usedFallback,
      available: true,
      error: errorMessage,
      fallback_available: true,
      details: { repository: repo, fallback_used: usedFallback },
    };

    return {
      source: 'github',
      events_count: events.length,
      sources_count: 1,
      summary_message: `Collected ${events.length} evidence events from 1 source (GitHub).`,
      events,
      connector_status: status,
      warnings,
    };
  }

  private async fetchCommits(repo: string, branch: string, since: string | undefined, token: string): Promise<any[]> {
    let url = `${this.apiBase}/repos/${repo}/commits?per_page=15&sha=${encodeURIComponent(branch)}`;
    if (since) url += `&since=${encodeURIComponent(since)}`;

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    const resp = await fetch(url, {
      headers: {
        Authorization: `Bearer ${token}`,
        'User-Agent': 'BlackBox-Incident-Investigator/1.0',
        Accept: 'application/vnd.github+json',
      },
      signal: controller.signal,
    });
    clearTimeout(timeout);

    if (!resp.ok) {
      throw new Error(`HTTP ${resp.status}: ${resp.statusText}`);
    }
    return await resp.json();
  }

  private async fetchPullRequests(repo: string, token: string): Promise<any[]> {
    const url = `${this.apiBase}/repos/${repo}/pulls?state=closed&sort=updated&per_page=5`;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    const resp = await fetch(url, {
      headers: {
        Authorization: `Bearer ${token}`,
        'User-Agent': 'BlackBox-Incident-Investigator/1.0',
        Accept: 'application/vnd.github+json',
      },
      signal: controller.signal,
    });
    clearTimeout(timeout);

    if (!resp.ok) return [];
    return await resp.json();
  }

  private async fetchDeployments(repo: string, token: string): Promise<any[]> {
    const url = `${this.apiBase}/repos/${repo}/deployments?per_page=5`;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    const resp = await fetch(url, {
      headers: {
        Authorization: `Bearer ${token}`,
        'User-Agent': 'BlackBox-Incident-Investigator/1.0',
        Accept: 'application/vnd.github+json',
      },
      signal: controller.signal,
    });
    clearTimeout(timeout);

    if (!resp.ok) return [];
    return await resp.json();
  }

  private normalizeCommit(raw: any, repo: string): EvidenceEvent | null {
    const sha = raw.sha || '';
    const commit = raw.commit || {};
    const message = commit.message || 'Commit';
    const headline = message.split('\n')[0];
    const author = commit.author?.name || raw.author?.login || 'Unknown';
    const timestamp = commit.author?.date || new Date().toISOString();
    const htmlUrl = raw.html_url;

    const isSuspicious = /pool|leak|revert|hotfix|timeout|perf|cache|concurrency|urgent/i.test(headline);

    return {
      id: `evt-gh-${sha.slice(0, 7)}`,
      source: 'github',
      event_type: 'commit',
      timestamp,
      title: `Commit [${sha.slice(0, 7)}]: ${headline}`,
      message,
      summary: `Git commit to ${repo} by ${author}: '${headline}'`,
      severity: isSuspicious ? 'warning' : 'info',
      metadata: {
        sha,
        author,
        repository: repo,
        commit_url: htmlUrl,
      },
      related_identifiers: [sha.slice(0, 7), author, repo],
      raw_reference: htmlUrl,
      correlated_event_ids: [],
      is_root_cause_candidate: isSuspicious,
    };
  }

  private normalizePullRequest(raw: any, repo: string): EvidenceEvent | null {
    const prNumber = raw.number;
    const title = raw.title || '';
    const body = raw.body || 'No description';
    const mergedAt = raw.merged_at;
    if (!mergedAt) return null;

    const author = raw.user?.login || 'unknown';
    const htmlUrl = raw.html_url;
    const isSuspicious = /sync|pool|refactor|bulk|migration|concurrency|worker/i.test(title);

    return {
      id: `evt-gh-pr-${prNumber}`,
      source: 'github',
      event_type: 'pull_request',
      timestamp: mergedAt,
      title: `PR #${prNumber} Merged: '${title}'`,
      message: body,
      summary: `Pull Request #${prNumber} by @${author} merged into main for ${repo}.`,
      severity: isSuspicious ? 'warning' : 'info',
      metadata: {
        pr_number: prNumber,
        author,
        merged_at: mergedAt,
        repository: repo,
        pr_url: htmlUrl,
      },
      related_identifiers: [`PR#${prNumber}`, `@${author}`, repo],
      raw_reference: htmlUrl,
      correlated_event_ids: [],
      is_root_cause_candidate: isSuspicious,
    };
  }

  private normalizeDeployment(raw: any, repo: string): EvidenceEvent | null {
    const depId = raw.id;
    const createdAt = raw.created_at || new Date().toISOString();
    const env = raw.environment || 'production';
    const ref = raw.ref || 'main';
    const creator = raw.creator?.login || 'ci-bot';

    return {
      id: `evt-gh-deploy-${depId}`,
      source: 'deployment',
      event_type: 'deployment',
      timestamp: createdAt,
      title: `GitHub Deployment to ${env} (${ref})`,
      message: `Automated deployment triggered by ${creator} for ref ${ref}.`,
      summary: `Deployment ${depId} promoted to ${env} for repository ${repo}.`,
      severity: 'info',
      metadata: {
        deployment_id: depId,
        environment: env,
        ref,
        creator,
        repository: repo,
      },
      related_identifiers: [String(depId), env, ref],
      raw_reference: raw.url,
      correlated_event_ids: [],
      is_root_cause_candidate: false,
    };
  }

  private async generateFallbackEvents(repo: string): Promise<EvidenceEvent[]> {
    const demo = new DemoProvider();
    const res = await demo.collectEvidence({ filter_source: 'github' });
    return res.events.map((e) => ({
      ...e,
      metadata: {
        ...e.metadata,
        is_fallback: true,
        fallback_repository: repo,
      },
    }));
  }
}
